<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">

<html>
		<head>
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="chrome=1">
		<title>vArmyKnife: Variant Army Knife Toolkit</title>

		<link rel="stylesheet" href="stylesheets/styles.2.css">
		<link rel="stylesheet" href="stylesheets/pygment_trac.css">
		<script src="javascripts/scale.fix.js"></script>
		<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

		<!--[if lt IE 9]>
		<script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
		<![endif]-->
</head>

<body>
<div class="wrapper">


<header>
			</section> <h1><a href="index.html">vArmyKnife</a></h1> <section>
			<p>Variant Army Knife</p>
</header>


<section>
</section> <h2>VARIANT OPERATIONS</h2> <section>

<pre><code>Variant Operations or Variant Map Functions are a set of 
    sub-utilities that perform operations on a variant set one 
    variant at a time. When more than one function is specified 
    in a run, these functions are performed in the order that 
    they appear in the command line, after all other operations 
    have been carried out (excluding output ops).
Basic Syntax:
    --FCN functionName|ID|param1=p1|param2=p2|...
</code></pre>

</section> <h2>Available Functions:</h2> <section>

<h3>snpEffExtract</h3>

<blockquote>
  
</blockquote>

<pre><code>annTag: A valid ANN formatted field, usually generated by 
    SNPeff.(String, default=ANN)
bioTypeKeepList: A comma delimited list of the transcript 
    biotypes that you want to keep. All other biotypes will be 
    ignored.(String)
effectKeepList: A comma delimited list of the effect types that 
    you want to keep. All other EFFECT values will be 
    ignored.(String)
warningDropList: A comma delimited list of warnings. Any 
    entries that include a listed warning will be 
    ignored.(String)
geneListName: (String)
geneList: (String)
severityList: Must be a list of severity levels, listed as some 
    combination of effectseverity types delimited with slashes. 
    Legal types are: HIGH, MODERATE, LOW, and MODIFIER, which 
    are standard SnpEFF effect types, and also: NS (HIGH and 
    MODERATE), NonNS (LOW and MODIFIER), and ANY (any 
    type).(String, default=HIGH/MODERATE/LOW)
extractFields: This is a complex multi-part field that allows 
    flexible extraction of information from SnpEff ANN tags. 
    This field must be in the colon-delimited format 
    tagInfix:ANN_idx:description:severityList[:noCollapse]. 
    severityList must be in the same format as the severityList 
    parameter above, but can override the default if desired. 
    ANN_idx must be a slash-delimited list of field indices 
    counting from zero in the ANN tag. The standard ANN field 
    indices are: 0:allele,1:effect,2:impact,3:geneName,4:geneID-
    ,5:txType,6:txID,7:txBiotype,8:rank,9:HGVS.c,10:HGVS.p,11:c-
    DNAposition,12:cdsPosition,13:proteinPosition,14:distToFeat-
    ure,15:warnings,16:errors If multiple fields are selected 
    then the output fields will have the format 
    first:second:third:etc. For example, to create two new 
    fields containing a list of all genes for which the current 
    variant has HIGH and MODERATE impact respectively, use the 
    format: myNewField:4:my description:HIGH/MODERATE. This 
    will generate two new fields: myNewField_HIGH and 
    myNewField_MODERATE. Note that if this function as a whole 
    has a mapID set, then both field names will be prefixed by 
    that overall ID.(String)
geneNameIdx: (String, default=4)
biotypeIdx: (String, default=7)
warnIdx: (String, default=15)
</code></pre>

<h3>fixSwappedRefAlt</h3>

<blockquote>
  <p>...</p>
</blockquote>

<pre><code>genomeFA: The genome fasta file containing the reference 
    genome. This will be used by various functions that require 
    genomic information. Note that some functions that call the 
    GATK library will also require that the fasta be indexed. 
    Note: Chromosome names must match.(String, required)
</code></pre>

<h3>convertSampleNames</h3>

<blockquote>
  <p>This function converts the sample IDs of the VCF file according to a decoder file that you supply. The decoder will decode </p>
</blockquote>

<pre><code>file: A tab delimited file with the from/to chromosome 
    names.(String, required)
columnNames: The column titles for the old chrom names and the 
    new chrom names, in that order. If this parameter is used, 
    the decoder file must have a title line.(String)
columnIdx: The column number of the current chromosome names 
    then the new chromosome names, in that order. Column 
    indices start counting from 0. If you use this parameter to 
    set the columns, and if the file has a title line, then you 
    should use skipFirstRow or else it will be read in as if it 
    were a chromosome.(Integer)
skipFirstRow: If this parameter is set, then this tool will 
    skip the first line on the decoder file. This is useful if 
    you are specifying the columns using column numbers but the 
    file also has a title line.(Flag)
</code></pre>

<h3>tagBedFile</h3>

<blockquote>
  <p>This function takes a BED file (which can be gzipped if desired) and creates a new INFO field based on whether the variant locus overlaps with a genomic region in the BED file. The new field can be either an integer that is equal to 1 if there is overlap and 0 otherwise (which is the default behavior) Or, alternatively, it can copy in the title field from the bed file. NOTE: this function only uses the first 3 to 5 fields of the BED file, it does not implement the optional fields 10-12 which can specify intron/exon blocks.</p>
</blockquote>

<pre><code>file: (String, required)
desc: (String, default=No desc provided)
buffer: (Integer, default=0)
style: (String, default=+)
</code></pre>

<h3>copyColumnToInfo</h3>

<blockquote>
  
</blockquote>

<pre><code>columnID: (String, required)
</code></pre>

<h3>checkReferenceMatch</h3>

<blockquote>
  <p>This function compares the REF column to the genomic reference and makes sure that they actually match. If mismatches are found, a warning will be thrown. In addition, a new INFO field will be added to the VCF that will be a simple integer field that will equal 1 if and only if the REF matches the reference, and 0 otherwise.</p>
</blockquote>

<pre><code>genomeFA: The genome fasta file containing the reference 
    genome. This will be used by various functions that require 
    genomic information. Note that some functions that call the 
    GATK library will also require that the fasta be indexed. 
    Note: Chromosome names must match.(String, required)
</code></pre>

<h3>keepVariants</h3>

<blockquote>
  <p>This function drops variants based on a given true/false expression.</p>
</blockquote>

<pre><code>expr: (String)
</code></pre>

<h3>dropNullVariants</h3>

<blockquote>
  <p>This function drops all lines with no alt alleles ('.' in the ALT column), or lines where the ALT allele is identical to the REF. Note: you must split multiallelics first. See the 'splitMultiallelics' function.</p>
</blockquote>

<pre><code>(This function takes no parameters)
</code></pre>

<h3>fixFirstBaseMismatch</h3>

<blockquote>
  <p>...</p>
</blockquote>

<pre><code>windowSize: Sets the size of the sliding window used. Problems 
    may occur if you have variants longer than this window 
    size. Default is 200bp.(Int)
genomeFA: The genome fasta file containing the reference 
    genome. This will be used by various functions that require 
    genomic information. Note that some functions that call the 
    GATK library will also require that the fasta be indexed. 
    Note: Chromosome names must match.(String, required)
</code></pre>

<h3>depthStats</h3>

<blockquote>
  <p>This function calculates various statistics on total read depth and hetAB.</p>
</blockquote>

<pre><code>inputGT: The input genotype FORMAT field.(String)
inputAD: (String)
inputDP: (String)
restrictToGroup: (String)
groupFile: A tab-delimited file containing sample ID's and a 
    list of group IDs for each sample. See the --groupFile 
    parameter of walkVcf.(String)
superGroupList: See the --superGroupList parameter of 
    walkVcf.(String)
</code></pre>

<h3>snpSiftDbnsfp</h3>

<blockquote>
  <p>This function runs the SnpSift dbnsfp command</p>
</blockquote>

<pre><code>cmd: A valid SnpSift command(String)
</code></pre>

<h3>calcBurdenMatrix</h3>

<blockquote>
  <p>....</p>
</blockquote>

<pre><code>geneTag: (String, required)
expr: (String)
sampleSet: (String)
group: (String)
inputGT: (String)
geneList: (String)
geneListFile: (String)
printFullGeneList: (flag)
pathwayList: (String)
groupFile: A tab-delimited file containing sample ID's and a 
    list of group IDs for each sample. See the --groupFile 
    parameter of walkVcf.(String)
superGroupList: See the --superGroupList parameter of 
    walkVcf.(String)
outfile: The output matrix file path.(String, required)
</code></pre>

<h3>splitMultiAllelics</h3>

<blockquote>
  <p>...</p>
</blockquote>

<pre><code>(This function takes no parameters)
</code></pre>

<h3>filterTags</h3>

<blockquote>
  <p>This function can be used to remove unwanted INFO or FORMAT fields, or remove unwanted samples. This can substantially reduce file sizes.</p>
</blockquote>

<pre><code>FORMAT.keep: (String)
FORMAT.drop: (String)
INFO.keep: (String)
INFO.drop: (String)
SAMPLES.keep: (String)
SAMPLES.drop: (String)
</code></pre>

<h3>gcContext</h3>

<blockquote>
  <p>This function calculates the fraction of bases within k bases from the variant locus that are G or C. This can be useful to identify high-GC areas where variant calling and sequencing may be less accurate.</p>
</blockquote>

<pre><code>windowSize: The number of bases to include in the context 
    window for determining local gc content.(String, required)
digits: Number of digits to round to.(String, default=4)
genomeFA: The genome fasta file containing the reference 
    genome. This will be used by various functions that require 
    genomic information. Note that some functions that call the 
    GATK library will also require that the fasta be indexed. 
    Note: Chromosome names must match.(String, required)
</code></pre>

<h3>tagVariantsExpression</h3>

<blockquote>
  <p>This function takes a variant expression and creates a new INFO field that is 1 if and only if that expression returns TRUE, and 0 otherwise.</p>
</blockquote>

<pre><code>expr: The variant expression, which is a true/false expression 
    using the variant expression syntax.(String, required)
desc: A description, to go in the info field 
    description.(String, default=No desc provided)
</code></pre>

<h3>tally</h3>

<blockquote>
  <p>This is a set of functions that takes various counts and totals across the whole VCF.</p>
</blockquote>

<pre><code>func: (String, required)
</code></pre>

<h3>addVariantPosInfo</h3>

<blockquote>
  <p>This function adds a new INFO field in the form: CHROM:START:REF>ALT. This can be useful for checking the effects of functions that alter the variant columns. For example, you can run this function before and after leftAlignAndTrim to see how a variant changes.</p>
</blockquote>

<pre><code>(This function takes no parameters)
</code></pre>

<h3>sanitize</h3>

<blockquote>
  
</blockquote>

<pre><code>(This function takes no parameters)
</code></pre>

<h3>snpEff</h3>

<blockquote>
  <p>This function runs SnpEff by calling the SnpEff library internally. It uses version 4.3t.</p>
</blockquote>

<pre><code>cmd: A valid SnpSift command(String)
</code></pre>

<h3>getLocusDepthFromWig</h3>

<blockquote>
  <p>...</p>
</blockquote>

<pre><code>wigfile: (String)
desc: (String)
</code></pre>

<h3>tagVariantsFunction</h3>

<blockquote>
  <p>This is a set of functions that all take one or more input parameters and outputs one new INFO field.</p>
</blockquote>

<pre><code>func: (String, required)
desc: (String, default=No desc provided)
digits: (Integer)
params: (String)
</code></pre>

<h3>convertToStdVcf</h3>

<blockquote>
  <p>Certain utilities (eg GATK) do not allow certain optional features of the VCFv4.2 format standard (For example: additional tag-pairs in the INFO or FORMAT header lines). This function strips out this additional metadata.</p>
</blockquote>

<pre><code>(This function takes no parameters)
</code></pre>

<h3>snpEffExtractField</h3>

<blockquote>
  
</blockquote>

<pre><code>tagPrefix: (String, required)
outputTagPrefix: (String)
columns: (String)
desc: (String)
collapseUniques: (Flag)
tagSet: (String)
</code></pre>

<h3>snpSiftAnno</h3>

<blockquote>
  <p>This function runs a SnpSift anno command</p>
</blockquote>

<pre><code>cmd: A valid SnpSift command(String)
</code></pre>

<h3>unPhaseAndSortGenotypes</h3>

<blockquote>
  <p>This function removes phasing and sorts genotypes (so that heterozygotes are always listed as 0/1 and never 1/0).</p>
</blockquote>

<pre><code>inputGT: The input/output genotype FORMAT field.(String)
groupFile: A tab-delimited file containing sample ID's and a 
    list of group IDs for each sample. See the --groupFile 
    parameter of walkVcf.(String)
superGroupList: See the --superGroupList parameter of 
    walkVcf.(String)
</code></pre>

<h3>calcStats</h3>

<blockquote>
  <p>This function combines the functions sampleCounts, sampleLists, and depthStats, performing all three.</p>
</blockquote>

<pre><code>groupFile: A tab-delimited file containing sample ID's and a 
    list of group IDs for each sample. See the --groupFile 
    parameter of walkVcf.(String)
superGroupList: See the --superGroupList parameter of 
    walkVcf.(String)
inputGT: The input genotype FORMAT field.(String)
inputAD: (String)
inputDP: ()
noCountsCalc: (Flag)
noFreqCalc: (Flag)
noMissCalc: (Flag)
noAlleCalc: (Flag)
noHetHomCalc: (Flag)
noMultiHetCalc: (Flag)
samplePrintLimit: (String)
noDepthStats: (Flag)
noSampleLists: (Flag)
noSampleCounts: (Flag)
expr: The variant expression, which is a true/false expression 
    using the variant expression syntax.(String)
</code></pre>

<h3>calcBurdenCounts</h3>

<blockquote>
  
</blockquote>

<pre><code>geneTag: (String, required)
expr: (String)
sampleSet: (String)
group: (String)
inputGT: (String)
groupFile: A tab-delimited file containing sample ID's and a 
    list of group IDs for each sample. See the --groupFile 
    parameter of walkVcf.(String)
superGroupList: See the --superGroupList parameter of 
    walkVcf.(String)
countFileID: If multiple output count files are desired, you 
    can specify which functions output to which count file 
    using this parameter. Note that each file must be created 
    using a --burdenCountsFile parameter, with the form 
    fileID:/path/to/file.txt(String)
</code></pre>

<h3>markDup</h3>

<blockquote>
  <p>This map function will detect duplicate variant lines and add two new INFO fields: mapID_CT and mapID_IDX. The CT will indicate how many duplicates were found matching the current variant, and the IDX will number each duplicate with a unique identifier, counting from 0. All nonduplicates will be marked with CT=1 and IDX=0.</p>
</blockquote>

<pre><code>(This function takes no parameters)
</code></pre>

<h3>homopolymerRunStats</h3>

<blockquote>
  
</blockquote>

<pre><code>runSize: The number of repeated bases required to count as a 
    homopolymer run(String, required)
genomeFA: The genome fasta file containing the reference 
    genome. This will be used by various functions that require 
    genomic information. Note that some functions that call the 
    GATK library will also require that the fasta be indexed. 
    Note: Chromosome names must match.(String, required)
</code></pre>

<h3>addDummyGenotypeColumn</h3>

<blockquote>
  
</blockquote>

<pre><code>(This function takes no parameters)
</code></pre>

<h3>concordanceCaller</h3>

<blockquote>
  <p>....</p>
</blockquote>

<pre><code>callerNames: Comma delimited list of caller IDs, used in the 
    callerSet INFO fields and the names of the output GT 
    fields. By default, callers will simply be named 
    C1,C2,...(String)
priority: Comma delimited list of caller IDs. The list of 
    caller IDs in order of descending priority.(String)
gtDecisionMethod: The merge rule for calculating 
    ensemble-merged GT and AD tags. Valid options are priority, 
    prioritySkipMissing, and majority_priorityOnTies. Default 
    is simple priority.(String)
ignoreSampleIds: If this flag is set, then sample IDs will be 
    ignored and each VCF will be assumed to have the exact same 
    samples in the exact same order. Use at your own risk.(flag)
ignoreSampleOrder: If this flag is set, then the sample IDs 
    will be used to match up the different VCFs, and the 
    samples may be in different orders in the different 
    files.(flag)
</code></pre>

<h3>addContextBases</h3>

<blockquote>
  <p>This function adds several new columns, </p>
</blockquote>

<pre><code>windowSize: The number of bases to include in the context 
    window(String, required)
genomeFA: The genome fasta file containing the reference 
    genome. This will be used by various functions that require 
    genomic information. Note that some functions that call the 
    GATK library will also require that the fasta be indexed. 
    Note: Chromosome names must match.(String, required)
</code></pre>

<h3>convertChromNames</h3>

<blockquote>
  <p>This function takes a file and translates chromosome names into a different format. This is most often used to convert between the chr1,chr2,... format and the 1,2,... format.</p>
</blockquote>

<pre><code>file: A tab delimited file with the from/to chromosome 
    names.(String, required)
columnNames: The column titles for the old chrom names and the 
    new chrom names, in that order. If this parameter is used, 
    the decoder file must have a title line.(String)
columnIdx: The column number of the current chromosome names 
    then the new chromosome names, in that order. Column 
    indices start counting from 0. If you use this parameter to 
    set the columns, and if the file has a title line, then you 
    should use skipFirstRow or else it will be read in as if it 
    were a chromosome.(Integer)
skipFirstRow: If this parameter is set, then this tool will 
    skip the first line on the decoder file. This is useful if 
    you are specifying the columns using column numbers but the 
    file also has a title line.(Flag)
</code></pre>

<h3>addVariantIdx</h3>

<blockquote>
  <p>This function adds a new INFO column with a unique numeric value for each line. Optionally, you can add a prefix to each ID.</p>
</blockquote>

<pre><code>prefix: Prefix to prepend to the index field.(String)
</code></pre>

<h3>dropSpanIndels</h3>

<blockquote>
  <p>This function drops Spanning indel lines ('*' alleles). Note: you must split multiallelics first!</p>
</blockquote>

<pre><code>(This function takes no parameters)
</code></pre>

<h3>addAltSequence</h3>

<blockquote>
  
</blockquote>

<pre><code>windowSize: The number of flanking bases to include on each 
    side of the alt sequence.(String, default=10)
genomeFA: The genome fasta file containing the reference 
    genome. This will be used by various functions that require 
    genomic information. Note that some functions that call the 
    GATK library will also require that the fasta be indexed. 
    Note: Chromosome names must match.(String, required)
</code></pre>

<h3>sampleLists</h3>

<blockquote>
  <p>This function generates sample list fields which contain comma-delimited lists of samples that are het or hom-alt.</p>
</blockquote>

<pre><code>inputGT: The input genotype FORMAT field.(String)
samplePrintLimit: (String)
groupFile: A tab-delimited file containing sample ID's and a 
    list of group IDs for each sample. See the --groupFile 
    parameter of walkVcf.(String)
superGroupList: See the --superGroupList parameter of 
    walkVcf.(String)
expr: The variant expression, which is a true/false expression 
    using the variant expression syntax.(String)
</code></pre>

<h3>fixDotAltIndels</h3>

<blockquote>
  <p>...</p>
</blockquote>

<pre><code>genomeFA: The genome fasta file containing the reference 
    genome. This will be used by various functions that require 
    genomic information. Note that some functions that call the 
    GATK library will also require that the fasta be indexed. 
    Note: Chromosome names must match.(String, required)
</code></pre>

<h3>dropGenotypeData</h3>

<blockquote>
  
</blockquote>

<pre><code>(This function takes no parameters)
</code></pre>

<h3>extractRegion</h3>

<blockquote>
  <p>This function extracts a single region from the VCF. NOTE: the VCF MUST BE SORTED!</p>
</blockquote>

<pre><code>region: The genomic region to extract.(String, required)
windowSize: The size of the window around the genomic region to 
    extract.(Int)
</code></pre>

<h3>addInfo</h3>

<blockquote>
  <p>This is a set of functions that all take one or more input parameters and outputs one new INFO field. The syntax is: --fcn "addInfo|newTagName|fcn(param1,param2,...)". Optionally you can add "|desc=tag description". There are numerous addInfo functions. For more information, go to the section on addInfo Functions below, or use the help command: varmyknife help addInfo</p>
</blockquote>

<pre><code>func: (String, required)
desc: The description in the header line for the new INFO 
    field.(String, default=No desc provided)
digits: For floating point values, the number of digits to 
    include after the decimal(Integer)
</code></pre>

<h3>calcBurdenCountsByGroups</h3>

<blockquote>
  
</blockquote>

<pre><code>geneTag: (String, required)
groups: (String, required)
expr: (String)
sampleSet: (String)
inputGT: (String)
groupFile: A tab-delimited file containing sample ID's and a 
    list of group IDs for each sample. See the --groupFile 
    parameter of walkVcf.(String)
superGroupList: See the --superGroupList parameter of 
    walkVcf.(String)
countFileID: If multiple output count files are desired, you 
    can specify which functions output to which count file 
    using this parameter. Note that each file must be created 
    using a --burdenCountsFile parameter, with the form 
    fileID:/path/to/file.txt(String)
</code></pre>

<h3>rmDup</h3>

<blockquote>
  <p>...</p>
</blockquote>

<pre><code>(This function takes no parameters)
</code></pre>

<h3>leftAlignAndTrim</h3>

<blockquote>
  <p>...</p>
</blockquote>

<pre><code>windowSize: Sets the size of the sliding window used. Problems 
    may occur if you have variants longer than this window 
    size. Default is 200bp.(Int)
genomeFA: The genome fasta file containing the reference 
    genome. This will be used by various functions that require 
    genomic information. Note that some functions that call the 
    GATK library will also require that the fasta be indexed. 
    Note: Chromosome names must match.(String, required)
</code></pre>

<h3>sampleCounts</h3>

<blockquote>
  <p>This function generates counts and frequencies for alt alleles, alt genotypes, missing genotypes, ref genotypes, and so on. Note that there are several calc- flags. If none of these are included, then this function does nothing.</p>
</blockquote>

<pre><code>groupFile: A tab-delimited file containing sample ID's and a 
    list of group IDs for each sample. See the --groupFile 
    parameter of walkVcf.(String)
superGroupList: See the --superGroupList parameter of 
    walkVcf.(String)
inputGT: The input genotype FORMAT field.(String)
noCountsCalc: (Flag)
noFreqCalc: (Flag)
noMissCalc: (Flag)
noAlleCalc: (Flag)
noHetHomCalc: (Flag)
noMultiHetCalc: (Flag)
</code></pre>

<h3>dropSymbolicAlleles</h3>

<blockquote>
  
</blockquote>

<pre><code>(This function takes no parameters)
</code></pre>

<h3>sampleReorder</h3>

<blockquote>
  <p>This function allows you to reorder the sample columns in your VCF. Set ONE of the parameters below to specify the desired ordering.</p>
</blockquote>

<pre><code>sampleOrdering: A simple list of all the samples, in the 
    desired order.(String)
sampleOrderingFile: A file containing one sampleID per line. 
    The samples will be reordered to match the order found in 
    the file.(String, required)
alphabetical: If this flag is set, then the samples will be 
    reordered alphabetically.(Flag)
</code></pre>

<h3>genotypeFilter</h3>

<blockquote>
  <p>This function filters a genotype field based on a given genotype expression. The new filtered genotype can replace the GT field or can be set to a different field, so multiple filtering strategies can be included in a single VCF.</p>
</blockquote>

<pre><code>expr: A Genotype Expression, using the genotype expression 
    syntax.(String, required)
desc: A description, to go in the new FORMAT fields.(String)
filterTag: The name of a new FORMAT field, which will be a flag 
    equal to 1 if and only if the genotype passes the 
    filter.(String)
outputGT: The output genotype FORMAT field. If this is the same 
    as the input genotype field then the genotype field will be 
    overwritten.(String)
inputGT: The input genotype FORMAT field.(String)
inputGtNewName: If this parameter is set, the input genotype 
    field will be copied to a new tag with this name before 
    filtering. This can be useful if overwriting the input 
    genotype field.()
groupFile: A tab-delimited file containing sample ID's and a 
    list of group IDs for each sample. See the --groupFile 
    parameter of walkVcf.(String)
superGroupList: See the --superGroupList parameter of 
    walkVcf.(String)
</code></pre>

</div></body></html>
