<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">

<html>
		<head>
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="chrome=1">
		<title>vArmyKnife: Variant Army Knife Toolkit</title>

		<link rel="stylesheet" href="stylesheets/styles.2.css">
		<link rel="stylesheet" href="stylesheets/pygment_trac.css">
		<script src="javascripts/scale.fix.js"></script>
		<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

		<!--[if lt IE 9]>
		<script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
		<![endif]-->
</head>

<body>
<div class="wrapper">


<header>
			</section> <h1><a href="index.html">vArmyKnife</a></h1> <section>
			<p>Variant Army Knife</p>
</header>


<section>
</section> <h1>vArmyKnife</h1> <section>

<blockquote>
  <p>Version3.2.6 (Updated Wed Jan 27 15:16:00 EST 2021)</p>

<p>(<a href="secondaryCommands.html">back to help base</a>)</p>
</blockquote>

</section> <h2>General Help</h2> <section>

</section> <h2>DESCRIPTION:</h2> <section>

<p>NOTE: if you run into OutOfMemoryExceptions, try adding the java options: "-Xmx8G"</p>

</section> <h2>GENERAL SYNTAX:</h2> <section>

<pre><code>varmyknife [java_options]  [options] infile outfile
OR
varmyknife [java_options]  [options] infile - &gt; outfile
OR
varmyknife [java_options] --CMD commandName [options]
</code></pre>

</section> <h2>COMMANDS:</h2> <section>

<h3><a href="index.html">walkVcf</a></h3>

<blockquote>
  <p>This utility performs a series of transformations on an input VCF file and adds an array of informative tags.</p>
</blockquote>

<h3><a href="VcfToMatrix.html">VcfToMatrix</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="splitExomeIntoParts.html">splitExomeIntoParts</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

</section> <h2>EXPERIMENTAL COMMANDS:</h2> <section>

<p>These commands have not been fully tested and are not for production use. Documentation may be incomplete or nonexistant. Some are nonfunctional, or have been subsumed into other commands, or were only intended for internal testing purposes. Use at your own risk!</p>

<h3><a href="GenerateTranscriptAnnotation.html">GenerateTranscriptAnnotation</a></h3>

<blockquote>
  <p>This utility ... WARNING: THIS UTILITY IS AN UNTESTED BETA, AND MAY CONTAIN MAJOR FLAWS. NOT FOR PRODUCTION USE! USE AT YOUR OWN RISK!</p>
</blockquote>

<h3><a href="filterGenotypesByStat.html">filterGenotypesByStat</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="CreateVariantSampleTableV2.html">CreateVariantSampleTableV2</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="RemoveUnwantedFields.html">RemoveUnwantedFields</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="FilterVCF.html">FilterVCF</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="legacyCanonInfo.html">legacyCanonInfo</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="CodingCoverageStats.html">CodingCoverageStats</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="addGroupSummaries.html">addGroupSummaries</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="addTxBed.html">addTxBed</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="CmdZip.html">CmdZip</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="runVcfTest.html">runVcfTest</a></h3>

<blockquote>
  <p>Test utility.</p>
</blockquote>

<h3><a href="legacyRedoDbnsfp.html">legacyRedoDbnsfp</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="legacyAddDomainInfo.html">legacyAddDomainInfo</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="testTXSeqUtil.html">testTXSeqUtil</a></h3>

<blockquote>
  <p>Test utility.</p>
</blockquote>

<h3><a href="CmdAssessACMG.html">CmdAssessACMG</a></h3>

<blockquote>
  <p>BETA: This function consolidates information from a wide variety of different input files and attempts to calculate a subset of the ACMG guidelines criteria. It then attempts to assign pathogenicity scores. WARNING: THIS UTILITY IS AN UNTESTED BETA, AND MAY CONTAIN MAJOR FLAWS. NOT FOR PRODUCTION USE! USE AT YOUR OWN RISK!</p>
</blockquote>

<h3><a href="legacySplit.html">legacySplit</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="GerpTrackToSpans.html">GerpTrackToSpans</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="ConvertToStandardVcf.html">ConvertToStandardVcf</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="MergeManyVcf.html">MergeManyVcf</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="ConvertChromNames.html">ConvertChromNames</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="buildSummaryTracks.html">buildSummaryTracks</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="oldWalkVcf.html">oldWalkVcf</a></h3>

<blockquote>
  <p>This utility performs a series of transformations on an input VCF file and adds an array of informative tags.</p>
</blockquote>

<h3><a href="legacyTxAnno.html">legacyTxAnno</a></h3>

<blockquote>
  <p>This utility adds an array of new VCF tags with information about the transcriptional changes caused by each variant. WARNING: THIS UTILITY IS AN UNTESTED BETA, AND MAY CONTAIN MAJOR FLAWS. NOT FOR PRODUCTION USE! USE AT YOUR OWN RISK!</p>
</blockquote>

<h3><a href="ConvertAminoRangeToGenoRange.html">ConvertAminoRangeToGenoRange</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="CalcVariantCountSummary.html">CalcVariantCountSummary</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="CalcGenotypeStatTable.html">CalcGenotypeStatTable</a></h3>

<blockquote>
  <p>This takes a stat table for genotype statistics (such as GQ or AD). Warning: does not function on phased genotypes! UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="fixVcfInfoWhitespace.html">fixVcfInfoWhitespace</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="SimSeqError.html">SimSeqError</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="compareVcfs.html">compareVcfs</a></h3>

<blockquote>
  <p>Compares two different VCFs containing different builds with overlapping sample sets.UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="VcfParserTest.html">VcfParserTest</a></h3>

<blockquote>
  <p>This utility adds an array of new VCF tags with information about the transcriptional changes caused by each variant. WARNING: THIS UTILITY IS AN UNTESTED BETA, AND MAY CONTAIN MAJOR FLAWS. NOT FOR PRODUCTION USE! USE AT YOUR OWN RISK!</p>
</blockquote>

<h3><a href="getHomopolymerBed.html">getHomopolymerBed</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="HelloWorld.html">HelloWorld</a></h3>

<blockquote>
  <p>Test utility that just says hello world.</p>
</blockquote>

<h3><a href="RecodeClinVarCLN.html">RecodeClinVarCLN</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="CreateVariantSampleTable.html">CreateVariantSampleTable</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="ibdSimulator.html">ibdSimulator</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

<h3><a href="extractSingletons.html">extractSingletons</a></h3>

<blockquote>
  <p>UNTESTED ALPHA: FOR DEVELOPMENT/TESTING PURPOSES ONLY! DO NOT USE!</p>
</blockquote>

</section> <h2>AUTHORS:</h2> <section>

<p>Stephen W. Hartley, Ph.D. stephen.hartley (at nih dot gov)</p>

</section> <h2>LEGAL:</h2> <section>

<pre><code>Written 2017-2019 by Stephen Hartley, PhD 
National Cancer Institute (NCI), Division of Cancer 
Epidemiology and Genetics (DCEG), Human Genetics Program
As a work of the United States Government, this software 
package and all related documentation and information is in the 
public domain within the United States. Additionally, the 
National Institutes of Health and the National Cancer Institute 
waives copyright and related rights in the work worldwide 
through the CC0 1.0 Universal Public Domain Dedication (which 
can be found at 
https://creativecommons.org/publicdomain/zero/1.0/).
Although all reasonable efforts have been taken to ensure the 
accuracy and reliability of the software and data, the National 
Human Genome Research Institute (NHGRI), the National Cancer 
Institute (NCI) and the U.S. Government does not and cannot 
warrant the performance or results that may be obtained by 
using this software or data. NHGRI, NCI and the U.S. Government 
disclaims all warranties as to performance, merchantability or 
fitness for any particular purpose.
In work or products derived from this material, proper 
attribution of the authors as the source of the software or 
data may be made using "NCI Division of Cancer Epidemiology and 
Genetics, Human Genetics Program" as the citation.
This package uses (but is not derived from) several 
externally-developed, open-source libraries which have been 
distributed under various open-source licenses. vArmyKnife is 
distributed with compiled versions of these packages.
Additional License information can be accessed using the 
command:
    vArmyKnife --help LICENSES
And can also found in the distributed source code in:
    src/main/resources/library.LICENSES.txt
</code></pre>

</div></body></html>
