<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">

<html>
		<head>
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="chrome=1">
		<title>vArmyKnife: Variant Army Knife Toolkit</title>

		<link rel="stylesheet" href="stylesheets/styles.2.css">
		<link rel="stylesheet" href="stylesheets/pygment_trac.css">
		<script src="javascripts/scale.fix.js"></script>
		<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

		<!--[if lt IE 9]>
		<script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
		<![endif]-->
</head>

<body>
<div class="wrapper">


<header>
			<h1><a href="index.html">vArmyKnife</a></h1>
			<p>Variant Army Knife</p>
</header>


<section>
<h1>VARIANT OPERATIONS</h1>

<pre><code>Variant Operations or Variant Map Functions are a set of 
sub\-utilities that perform operations on a variant set one 
variant at a time\. When more than one function is specified in 
a run, these functions are performed in the order that they 
appear in the command line, after all other operations have 
been carried out \(excluding output ops\)\.
Basic Syntax:
    \-\-FCN functionName|ID|param1=p1|param2=p2|\.\.\.
</code></pre>

</section> <h2>Available Functions:</h2> <section>

<h3>snpEffExtract</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
annTag: A valid SnpSift command\(String, default=ANN\)
bioTypeKeepList: \(String\)
effectKeepList: \(String\)
warningDropList: \(String\)
keepIdx: \(String\)
geneListName: \(String\)
geneList: \(String\)
</code></pre>

<h3>contextBases</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
windowSize: The number of bases to include in the context 
window\(String, required\)
genomeFA: A genome fasta file, if different from the global 
\-\-genomeFA parameter\(String\)
</code></pre>

<h3>keepVariants</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
expr: \(String\)
</code></pre>

<h3>depthStats</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
inputGT: The input genotype FORMAT field\.\(String\)
inputAD: \(String\)
inputDP: \(String\)
restrictToGroup: \(String\)
groupFile: A tab\-delimited file containing sample ID's and a 
list of group IDs for each sample\. See the \-\-groupFile 
parameter of walkVcf\.\(String\)
superGroupList: See the \-\-superGroupList parameter of 
walkVcf\.\(String\)
</code></pre>

<h3>snpSiftDbnsfp</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
cmd: A valid SnpSift command\(String\)
</code></pre>

<h3>filterTags</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
FORMAT\.keep: \(String\)
FORMAT\.drop: \(String\)
INFO\.keep: \(String\)
INFO\.drop: \(String\)
SAMPLES\.keep: \(String\)
SAMPLES\.drop: \(String\)
</code></pre>

<h3>gcContext</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
windowSize: The number of bases to include in the context 
window for determining local gc content\.\(String, required\)
digits: Number of digits to round to\.\(String, default=4\)
genomeFA: A genome fasta file, if different from the global 
\-\-genomeFA parameter\(String\)
</code></pre>

<h3>tagVariantsExpression</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
expr: The variant expression, which is a true/false expression 
using the variant expression syntax\.\(String, required\)
desc: A description, to go in the info field 
description\.\(String, default=No desc provided\)
</code></pre>

<h3>snpEff</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
cmd: A valid SnpSift command\(String\)
</code></pre>

<h3>tagVariantsFunction</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
func: \(String, required\)
desc: \(String, default=No desc provided\)
digits: \(Integer\)
params: \(String\)
</code></pre>

<h3>snpEffExtractField</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
tagPrefix: \(String, required\)
outputTagPrefix: \(String\)
columns: \(String\)
desc: \(String\)
collapseUniques: \(Flag\)
tagSet: \(String\)
</code></pre>

<h3>snpSiftAnno</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
cmd: A valid SnpSift command\(String\)
</code></pre>

<h3>calcStats</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
groupFile: A tab\-delimited file containing sample ID's and a 
list of group IDs for each sample\. See the \-\-groupFile 
parameter of walkVcf\.\(String\)
superGroupList: See the \-\-superGroupList parameter of 
walkVcf\.\(String\)
inputGT: The input genotype FORMAT field\.\(String\)
calcCounts: \(Flag\)
calcFreq: \(Flag\)
calcMiss: \(Flag\)
calcAlle: \(Flag\)
calcHetHom: \(Flag\)
calcMultiHet: \(Flag\)
inputAD: \(String\)
inputDP: \(\)
calcCounts: \(Flag\)
calcFreq: \(Flag\)
calcMisc: \(Flag\)
calcAlle: \(Flag\)
calcHetHom: \(Flag\)
calcMultiHet: \(Flag\)
samplePrintLimit: \(String\)
noDepthStats: \(Flag\)
noSampleLists: \(Flag\)
noSampleCounts: \(Flag\)
expr: The variant expression, which is a true/false expression 
using the variant expression syntax\.\(String\)
</code></pre>

<h3>calcBurdenCounts</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
geneTag: \(String, required\)
snpEffAnnField: NOT IMPLEMENTED\(String\)
snpEffGeneField: NOT IMPLEMENTED\(String\)
snpEffVariantTypes: NOT IMPLEMENTED\(String\)
expr: \(String\)
sampleSet: \(String\)
inputGT: \(String\)
groupFile: A tab\-delimited file containing sample ID's and a 
list of group IDs for each sample\. See the \-\-groupFile 
parameter of walkVcf\.\(String\)
superGroupList: See the \-\-superGroupList parameter of 
walkVcf\.\(String\)
countFileID: If multiple output count files are desired, you 
can specify which functions output to which count file using 
this parameter\. Note that each file must be created using a 
\-\-burdenCountsFile parameter, with the form 
fileID:/path/to/file\.txt\(String\)
</code></pre>

<h3>homopolymerRunStats</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
runSize: The number of repeated bases required to count as a 
homopolymer run\(String, required\)
genomeFA: A genome fasta file, if different from the global 
\-\-genomeFA parameter\(String\)
</code></pre>

<h3>addVariantIdx</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
windowSize: The number of bases to include in the context 
window for determining local gc content\.\(String, required\)
genomeFA: A genome fasta file, if different from the global 
\-\-genomeFA parameter\(String\)
</code></pre>

<h3>sampleLists</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
inputGT: The input genotype FORMAT field\.\(String\)
samplePrintLimit: \(String\)
groupFile: A tab\-delimited file containing sample ID's and a 
list of group IDs for each sample\. See the \-\-groupFile 
parameter of walkVcf\.\(String\)
superGroupList: See the \-\-superGroupList parameter of 
walkVcf\.\(String\)
expr: The variant expression, which is a true/false expression 
using the variant expression syntax\.\(String\)
</code></pre>

<h3>sampleCounts</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
groupFile: A tab\-delimited file containing sample ID's and a 
list of group IDs for each sample\. See the \-\-groupFile 
parameter of walkVcf\.\(String\)
superGroupList: See the \-\-superGroupList parameter of 
walkVcf\.\(String\)
inputGT: The input genotype FORMAT field\.\(String\)
calcCounts: \(Flag\)
calcFreq: \(Flag\)
calcMiss: \(Flag\)
calcAlle: \(Flag\)
calcHetHom: \(Flag\)
calcMultiHet: \(Flag\)
</code></pre>

<h3>genotypeFilter</h3>

<pre><code>mapType: The function to run\.\(String, required\)
mapID: The ID to use for this operation\. This ID will also be 
used in the output tag IDs generated by this 
operation\.\(String, required\)
expr: A Genotype Expression, using the genotype expression 
syntax\.\(String, required\)
desc: A description, to go in the new FORMAT fields\.\(String\)
filterTag: The name of a new FORMAT field, which will be a flag 
equal to 1 if and only if the genotype passes the 
filter\.\(String\)
outputGT: The output genotype FORMAT field\. If this is the 
same as the input genotype field then the genotype field will 
be overwritten\.\(String\)
inputGT: The input genotype FORMAT field\.\(String\)
inputGtNewName: If this parameter is set, the input genotype 
field will be copied to a new tag with this name before 
filtering\. This can be useful if overwriting the input 
genotype field\.\(\)
groupFile: A tab\-delimited file containing sample ID's and a 
list of group IDs for each sample\. See the \-\-groupFile 
parameter of walkVcf\.\(String\)
superGroupList: See the \-\-superGroupList parameter of 
walkVcf\.\(String\)
</code></pre>

</div></body></html>
