
# VARIANT OPERATIONS

    
    Variant Operations or Variant Map Functions are a set of 
    sub\-utilities that perform operations on a variant set one 
    variant at a time\. When more than one function is specified in 
    a run, these functions are performed in the order that they 
    appear in the command line, after all other operations have 
    been carried out \(excluding output ops\)\.
    Basic Syntax:
        \-\-FCN functionName|ID|param1=p1|param2=p2|\.\.\.

## Available Functions:

    

### snpEffExtract

    
    
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    annTag: A valid SnpSift command\(String, default=ANN\)
    bioTypeKeepList: \(String\)
    effectKeepList: \(String\)
    warningDropList: \(String\)
    keepIdx: \(String\)
    geneListName: \(String\)
    geneList: \(String\)

### tagBedFile

    
    \.\.\.\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    file: \(String, required\)
    desc: \(String, default=No desc provided\)
    buffer: \(Integer, default=0\)
    style: \(String, default=\+\)

### checkReferenceMatch

    
    
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    genomeFA: \(String, required\)

### keepVariants

    
    This function drops variants based on a given true/false 
    expression\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    expr: \(String\)

### depthStats

    
    This function calculates various statistics on total read depth 
    and hetAB\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    inputGT: The input genotype FORMAT field\.\(String\)
    inputAD: \(String\)
    inputDP: \(String\)
    restrictToGroup: \(String\)
    groupFile: A tab\-delimited file containing sample ID's and a 
    list of group IDs for each sample\. See the \-\-groupFile 
    parameter of walkVcf\.\(String\)
    superGroupList: See the \-\-superGroupList parameter of 
    walkVcf\.\(String\)

### snpSiftDbnsfp

    
    This function runs the SnpSift dbnsfp command
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    cmd: A valid SnpSift command\(String\)

### calcBurdenMatrix

    
    \.\.\.\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    geneTag: \(String, required\)
    expr: \(String\)
    sampleSet: \(String\)
    group: \(String\)
    inputGT: \(String\)
    geneList: \(String\)
    geneListFile: \(String\)
    printFullGeneList: \(flag\)
    pathwayList: \(String\)
    groupFile: A tab\-delimited file containing sample ID's and a 
    list of group IDs for each sample\. See the \-\-groupFile 
    parameter of walkVcf\.\(String\)
    superGroupList: See the \-\-superGroupList parameter of 
    walkVcf\.\(String\)
    outfile: The output matrix file path\.\(String, required\)

### filterTags

    
    This function can be used to remove unwanted INFO or FORMAT 
    fields, or remove unwanted samples\. This can substantially 
    reduce file sizes\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    FORMAT\.keep: \(String\)
    FORMAT\.drop: \(String\)
    INFO\.keep: \(String\)
    INFO\.drop: \(String\)
    SAMPLES\.keep: \(String\)
    SAMPLES\.drop: \(String\)

### gcContext

    
    
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    windowSize: The number of bases to include in the context 
    window for determining local gc content\.\(String, required\)
    digits: Number of digits to round to\.\(String, default=4\)
    genomeFA: \(String, required\)

### tagVariantsExpression

    
    This function takes a variant expression and creates a new INFO 
    field that is 1 if and only if that expression returns TRUE, 
    and 0 otherwise\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    expr: The variant expression, which is a true/false expression 
    using the variant expression syntax\.\(String, required\)
    desc: A description, to go in the info field 
    description\.\(String, default=No desc provided\)

### sanitize

    
    
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)

### snpEff

    
    This function runs SnpEff
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    cmd: A valid SnpSift command\(String\)

### tagVariantsFunction

    
    This is a set of functions that all take one or more input 
    parameters and outputs one new INFO field\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    func: \(String, required\)
    desc: \(String, default=No desc provided\)
    digits: \(Integer\)
    params: \(String\)

### snpEffExtractField

    
    
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    tagPrefix: \(String, required\)
    outputTagPrefix: \(String\)
    columns: \(String\)
    desc: \(String\)
    collapseUniques: \(Flag\)
    tagSet: \(String\)

### snpSiftAnno

    
    This function runs a SnpSift anno command
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    cmd: A valid SnpSift command\(String\)

### calcStats

    
    This function combines the functions sampleCounts, sampleLists, 
    and depthStats, performing all three\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    groupFile: A tab\-delimited file containing sample ID's and a 
    list of group IDs for each sample\. See the \-\-groupFile 
    parameter of walkVcf\.\(String\)
    superGroupList: See the \-\-superGroupList parameter of 
    walkVcf\.\(String\)
    inputGT: The input genotype FORMAT field\.\(String\)
    inputAD: \(String\)
    inputDP: \(\)
    noCountsCalc: \(Flag\)
    noFreqCalc: \(Flag\)
    noMissCalc: \(Flag\)
    noAlleCalc: \(Flag\)
    noHetHomCalc: \(Flag\)
    noMultiHetCalc: \(Flag\)
    samplePrintLimit: \(String\)
    noDepthStats: \(Flag\)
    noSampleLists: \(Flag\)
    noSampleCounts: \(Flag\)
    expr: The variant expression, which is a true/false expression 
    using the variant expression syntax\.\(String\)

### calcBurdenCounts

    
    
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    geneTag: \(String, required\)
    snpEffAnnField: NOT IMPLEMENTED\(String\)
    snpEffGeneField: NOT IMPLEMENTED\(String\)
    snpEffVariantTypes: NOT IMPLEMENTED\(String\)
    expr: \(String\)
    sampleSet: \(String\)
    group: \(String\)
    inputGT: \(String\)
    groupFile: A tab\-delimited file containing sample ID's and a 
    list of group IDs for each sample\. See the \-\-groupFile 
    parameter of walkVcf\.\(String\)
    superGroupList: See the \-\-superGroupList parameter of 
    walkVcf\.\(String\)
    countFileID: If multiple output count files are desired, you 
    can specify which functions output to which count file using 
    this parameter\. Note that each file must be created using a 
    \-\-burdenCountsFile parameter, with the form 
    fileID:/path/to/file\.txt\(String\)

### homopolymerRunStats

    
    
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    runSize: The number of repeated bases required to count as a 
    homopolymer run\(String, required\)
    genomeFA: \(String, required\)

### addContextBases

    
    
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    windowSize: The number of bases to include in the context 
    window\(String, required\)
    genomeFA: \(String, required\)

### convertChromNames

    
    \.\.\.\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    file: A tab delimited file with the from/to chromosome 
    names\.\(String, required\)
    columnNames: The column titles for the old chrom names and the 
    new chrom names, in that order\. If this parameter is used, the 
    decoder file must have a title line\.\(String\)
    columnIdx: The column number of the current chromosome names 
    then the new chromosome names, in that order\. Column indices 
    start counting from 0\. If you use this parameter to set the 
    columns, and if the file has a title line, then you should use 
    skipFirstRow or else it will be read in as if it were a 
    chromosome\.\(Integer\)
    skipFirstRow: \(Flag\)

### addVariantIdx

    
    
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    windowSize: The number of bases to include in the context 
    window for determining local gc content\.\(String, required\)
    genomeFA: \(String, required\)

### sampleLists

    
    This function generates sample list fields which contain 
    comma\-delimited lists of samples that are het or hom\-alt\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    inputGT: The input genotype FORMAT field\.\(String\)
    samplePrintLimit: \(String\)
    groupFile: A tab\-delimited file containing sample ID's and a 
    list of group IDs for each sample\. See the \-\-groupFile 
    parameter of walkVcf\.\(String\)
    superGroupList: See the \-\-superGroupList parameter of 
    walkVcf\.\(String\)
    expr: The variant expression, which is a true/false expression 
    using the variant expression syntax\.\(String\)

### sampleCounts

    
    This function generates counts and frequencies for alt alleles, 
    alt genotypes, missing genotypes, ref genotypes, and so on\. 
    Note that there are several calc\- flags\. If none of these are 
    included, then this function does nothing\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    groupFile: A tab\-delimited file containing sample ID's and a 
    list of group IDs for each sample\. See the \-\-groupFile 
    parameter of walkVcf\.\(String\)
    superGroupList: See the \-\-superGroupList parameter of 
    walkVcf\.\(String\)
    inputGT: The input genotype FORMAT field\.\(String\)
    noCountsCalc: \(Flag\)
    noFreqCalc: \(Flag\)
    noMissCalc: \(Flag\)
    noAlleCalc: \(Flag\)
    noHetHomCalc: \(Flag\)
    noMultiHetCalc: \(Flag\)

### genotypeFilter

    
    This function filters a genotype field based on a given 
    genotype expression\. The new filtered genotype can replace the 
    GT field or can be set to a different field, so multiple 
    filtering strategies can be included in a single VCF\.
    mapType: The function to run\.\(String, required\)
    mapID: The ID to use for this operation\. This ID will also be 
    used in the output tag IDs generated by this 
    operation\.\(String, required\)
    expr: A Genotype Expression, using the genotype expression 
    syntax\.\(String, required\)
    desc: A description, to go in the new FORMAT fields\.\(String\)
    filterTag: The name of a new FORMAT field, which will be a flag 
    equal to 1 if and only if the genotype passes the 
    filter\.\(String\)
    outputGT: The output genotype FORMAT field\. If this is the 
    same as the input genotype field then the genotype field will 
    be overwritten\.\(String\)
    inputGT: The input genotype FORMAT field\.\(String\)
    inputGtNewName: If this parameter is set, the input genotype 
    field will be copied to a new tag with this name before 
    filtering\. This can be useful if overwriting the input 
    genotype field\.\(\)
    groupFile: A tab\-delimited file containing sample ID's and a 
    list of group IDs for each sample\. See the \-\-groupFile 
    parameter of walkVcf\.\(String\)
    superGroupList: See the \-\-superGroupList parameter of 
    walkVcf\.\(String\)