<!DOCTYPE html>

<html>
		<head>
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="chrome=1">
		<title>vArmyKnife: Variant Army Knife Toolkit</title>

		<link rel="stylesheet" href="docs/stylesheets/styles.css">
		<link rel="stylesheet" href="docs/stylesheets/pygment_trac.css">
		<script src="javascripts/scale.fix.js"></script>
		<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

		<!--[if lt IE 9]>
		<script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
		<![endif]-->
</head>

<body>
<div class="wrapper">


<header>
			<h1><a href="index.html">vArmyKnife</a></h1>
			<p>Variant Army Knife</p>
			<p class="view"><a href="https://github.com/hartleys/vArmyKnife">View the Project on GitHub <small>Stephen Hartley</small></a></p>
			<ul>
				<li><a href="https://github.com/hartleys/vArmyKnife/releases/latest">Latest <strong>Release</strong></a></li>
				<li><a href="https://github.com/hartleys/vArmyKnife/zipball/master">Download <strong>ZIP File</strong></a></li>
				<li><a href="https://github.com/hartleys/vArmyKnife/tarball/master">Download <strong>TAR Ball</strong></a></li>
				<li><a href="https://github.com/hartleys/vArmyKnife">View On <strong>GitHub</strong></a></li>
			</ul>
</header>


<section><h1>User Manual for vArmyKnife</h1>

<blockquote>
  <p>v2.1.84   \
Compiled Tue Aug 14 15:42:38 EDT 2018</p>
</blockquote>

</section> <h2>QUICK START:</h2> <section>

<p><a href="https://github.com/hartleys/vArmyKnife/releases">You can download the most recent stable version of vArmyKnife here</a>, 
or you can use the <a href="https://github.com/hartleys/vArmyKnife/tarball/master">most recent experimental build here</a>.</p>

<p>Simply download the vArmyKnife.tar.gz file and extract it to your preferred location.</p>

<pre><code>tar xvzf varmyknife.tar.gz /my/install/directory/
</code></pre>

<p>If you are running either Linux or OSX, you can install the software onto your PATH using the command:</p>

<pre><code>export PATH=/my/install/directory/:$PATH
</code></pre>

<p>Then you can test vArmyKnife and see the basic syntax using either of the commands:</p>

<pre><code>varmyknife help
   or
java -jar /my/install/directory/vArmyKnife.jar help
</code></pre>

</section> <h2>RECOMMENDED ENVIRONMENT AND JAVA OPTIONS:</h2> <section>

<p>I have found that the following environment variables seem to improve stability and performance
when using vArmyKnife on a cluster or HPC-like environment.</p>

<pre><code>export _JAVA_OPTIONS="-Xms1g -XX:ParallelGCThreads=1 -Xss512m -XX:+UseConcMarkSweepGC -XX:+CMSClassUnloadingEnabled"
export MALLOC_ARENA_MAX=1
</code></pre>

<p>Also, you're usually going to want to set the maximum memory. How much memory is required 
will depend on the type of job. The more large annotation files that need to be loaded,
the more memory will be required. Note however that if you restrict to only one chromosome,
most of the annotation files will only be loaded for that chromosome, greatly reducing the
memory footprint.</p>

<p>You can set the memory usage by adding "-Xmx4g" just after "varmyknife" when you invoke vArmyKnife,
changing "4g" to however many gigabytes you think you will need. Note that if java starts to 
run low on memory it will slow down considerably as it tries to conserve. It's usually best 
to allocate some extra memory as needed. So for example, to set the max memory to 4 gigabytes:</p>

<pre><code>varmyknife -Xmx4g walkVcf inputVcf.vcf.gz outputVcf.vcf.gz
</code></pre>

</section> <h2>COMPLETE DOCUMENTATION:</h2> <section>

<p>General command documentation can be found <a href="docs/index.html"><em>here</em></a>.</p>

<p>Options and variables for the walkVcf command (which is the command that you will almost always use for most purposes)
can be found <a href="walkVcf.html"><em>here</em></a></p>

</section> <h2>INPUT COMMANDS AND OPTIONS:</h2> <section>

<p>All vArmyKnife executions follow the same basic syntax:</p>

<pre><code>varmyknife commandName parameters
</code></pre>

<p>For example:</p>

<pre><code>varmyknife walkVcf testInput.vcf.gz testOutput.vcf.gz
</code></pre>

<p>All the primary functions of vArmyKnife have been folded into the multiStepPipeline command,
so you're almost always going to use that command. The above invocation is the simplest
possible use of this tool. It simply reads in a VCF file (which can be gzipped, or not) and
copies it to a new place (which can also be gzipped, or not).</p>

<p>You can add java options prior to the "command" part, and vArmyKnife options after the command. 
So for example:</p>

<pre><code>varmyknife -Xmx5g multiStepPipeline --verbose testInput.vcf.gz testOutput.vcf.gz
</code></pre>

<p>Will increase the allowed memory to 5 gigabytes and turn on verbose reporting (making it tell you
more progress and debugging information).</p>

<p>In addition to the mandatory inputVcf and outputVcf parameters, multiStepPipeline
can take a large number of additional optional parameters that add functionality to the
run. These options should be placed AFTER the multiStepPipeline command but before the mandatory parameters. 
For example, if you simply want to take an input VCF and split all multiallelics:</p>

<pre><code>varmyknife -Xmx5g multiStepPipeline --splitMultiAllelics testInput.vcf.gz testOutput.vcf.gz
</code></pre>

<p>You can get a full list of all parameters using the command:</p>

<pre><code>varmyknife multiStepPipeline --help
</code></pre>

</section> <h2>VCF Walkers:</h2> <section>

<p>The basic structure of the software was inspired by GATK. 
Different commands execute different "VCF walkers". </p>

<p>A "walker" is just a modular sub-program that 
takes a VCF, does something to it, and passes it on. </p>

<p>Unlike with GATK, however, my tool can "chain"
multiple walkers in sequence, allowing a large number of steps to be merged into one big step. 
This is very useful, because at least on CCAD and HELIX, the process of writing a VCF to disk and then 
reading it out again is very time consuming. In addition to reducing the number of steps required, 
using chained walkers greatly reduces the total runtime.</p>

<p>When using the "vArmyKnife multiStepPipeline" command, VCF walkers are added depending on 
what optional parameters you add. If all you include is the input and output files (which are required), 
then all it does is read out the VCF and then copy it over:</p>

<pre><code>varmyknife multiStepPipeline inputfile.vcf.gz outputfile.vcf.gz
</code></pre>

<p>But, if you add a GTF annotation file and a genome fasta file:</p>

<pre><code>varmyknife walkVcf --gtfFile myGtfFile.gtf.gz \
                   --genomeFA hg19.fa.gz \
                     testInput.vcf.gz \
                     testOutput.vcf.gz
</code></pre>

<p>Then it will add a "AddTxAnnoSVcfWalker" walker, which adds a wide array of transcript-related annotations, like 
transcript lists, mRNA changes, amino acid changes, variant type (LOF/splice-loss/etc), and so on.</p>

<p>If you also add a file that lists transcript names and the associated gene symbol, 
it will also include a gene list for each variant:</p>

<pre><code>varmyknife walkVcf --txToGene txToGene.txt \
                   --gtfFile myGtfFile.gtf.gz \
                   --genomeFA hg19.fa.gz \
                     testInput.vcf.gz \
                     testOutput.vcf.gz
</code></pre>

<p>If you also add a file that lists the canonical transcript for each gene, then it will add another walker
that generates similar annotation fields restricted to only canonical transcripts:</p>

<pre><code>varmyknife walkVcf --txToGene txToGene.txt \
                   --gtfFile myGtfFile.gtf.gz \
                   --genomeFA hg19.fa.gz \
                   --addCanonicalTags canonList.txt \
                     testInput.vcf.gz \
                     testOutput.vcf.gz
</code></pre>

<p>If you add the parameter --splitMultiAllelics, then it will also split any multiallelic variants:</p>

<pre><code>varmyknife walkVcf --txToGene txToGene.txt \
                   --gtfFile myGtfFile.gtf.gz \
                   --genomeFA hg19.fa.gz \
                   --addCanonicalTags canonList.txt \
                   --splitMultiAllelics \
                     testInput.vcf.gz \
                     testOutput.vcf.gz
</code></pre>

<p>If you also add the "--groupFile" parameter, specifying a file that lists all the samples along with
their sample group(s), then you will get various counts for each sample group:</p>

<pre><code>varmyknife walkVcf --txToGene txToGene.txt \
                   --gtfFile myGtfFile.gtf.gz \
                   --genomeFA hg19.fa.gz \
                   --addCanonicalTags canonList.txt \
                   --splitMultiAllelics \
                   --groupFile testGroups.txt \
                     testInput.vcf.gz \
                     testOutput.vcf.gz
</code></pre>

<p>Sometimes with all these parameters, the commands can get really long and complex. To help alleviate 
this, you can specify a paramFile:</p>

<pre><code>varmyknife walkVcf --paramFile testParam.part001.txt \
                            testInput.vcf.gz \
                            testOutput.vcf.gz
</code></pre>

<p>With the paramFile "testParam.part001.txt" that looks like this:</p>

<pre><code>--txToGene  txToGene.txt 
--groupFile testGroups.txt 
--inputSavedTxFile txdata.txt.gz
--addCanonicalTags canonList.txt
--splitMultiAllelics
</code></pre>

</section> <h2>Param files: advanced</h2> <section>

<p>The paramFile has a number of extra functions intended make it easier to use.
First of all, all leading whitespace on any line is automatically removed.
Secondly, any lines in a paramFile that (after trimming leading whitespace) begin with a "#", are ignored.
In addition, if lines end with backslashes, they are combined with the next line (with leading whitespace removed).</p>

<p>So, for example, this is a legal paramfile:</p>

<pre><code># Tally groups:
--groupFile            testGroups.txt 
# Add transcript stuff:
    --txToGene         ucsc/txToGene.txt 
    --inputSavedTxFile ucsc/txdata.txt.gz
    --addCanonicalTags ucsc/canonList.txt
#Extra stuff:
    --splitMultiAllelics
#Add repeatmasker mask:
    --addBedTags SWH_locusIsNearRmsk_5:\
                 5:+:\
                 Variant locus is within 5bp of a repeatmasker masked region:\
                 rmsk.bed.gz,\
</code></pre>

<p>Finally, you can also set internal variables that will be expanded out whenever they are invoked.
Variables are created using the syntax "varname=varText", and are invoked using the syntax "${varname}".</p>

<p>So, for example, this is a legal paramfile:</p>

<pre><code># Set to annotation directory:
RESOURCE_DIR=/set/me/to/resource/dir/
# Now, instead of having to change every variable's base directory, 
#      you only need to change the line above.
#      Note that advanced bash substitutions are NOT supported.
# You can also set further variables that call previous variables:
UCSCDATA_DIR=${RESOURCE_DIR}ucsc/
# Tally groups:
--groupFile            ${RESOURCE_DIR}testGroups.txt 
# Add transcript stuff:
    --txToGene         ${UCSCDATA_DIR}txToGene.txt 
    --inputSavedTxFile ${UCSCDATA_DIR}txdata.txt.gz
    --addCanonicalTags ${UCSCDATA_DIR}canonList.txt
#Extra stuff:
    --splitMultiAllelics
#Add repeatmasker mask:
    --addBedTags SWH_locusIsNearRmsk_5:\
                 5:+:\
                 Variant locus is within 5bp of a repeatmasker masked region:\
                 ${UCSCDATA_DIR}rmsk.bed.gz
</code></pre>

<p>Note that these internal variables are ONLY used INSIDE the paramfile. Any invocations in the parameters
outside the paramfile will NOT be substituted out (or might even get substituted out by the command line itself,
if you're using bash).</p>

</section> <h2>MERGING MULTIPLE CALLER VCFS:</h2> <section>

<p>TODO: write more!</p>

</section> <h2>OUTPUT DOCUMENTATION:</h2> <section>

<p>TODO: write more!</p>

</section> <h2>LEGAL:</h2> <section>

<p>Written 2017 by Stephen Hartley, PhD </p>

<p>National Cancer Institute (NCI), Division of Cancer Epidemiology and Genetics (DCEG), Human Genetics Program</p>

<p>vArmyKnife and all relevant documentation is "United States Government Work" under he terms of the United States Copyright Act. It was written as part of the authors' official duties for the United States Government and thus vArmyKnife cannot be copyrighted. This software is freely available to the public for use without a copyright notice. Restrictions cannot be placed on its present or future use.</p>

<p>Although all reasonable efforts have been taken to ensure the accuracy and reliability of the software and data, the National Human Genome Research Institute (NHGRI), the National Cancer Institute (NCI) and the U.S. Government does not and cannot warrant the performance or results that may be obtained by using this software or data. NHGRI, NCI and the U.S. Government disclaims all warranties as to performance, merchantability or fitness for any particular purpose.</p>

<p>In any work or product derived from this material, proper attribution of the authors as the source of the software or data should be made, using "NCI Division of Cancer Epidemiology and Genetics, Human Genetics Program" as the citation.</p>

<p>This package uses (but is not derived from) several externally-developed, open-source libraries which have been distributed under various open-source licenses. vArmyKnife is distributed packaged with these libraries included.</p>

<p>Additional License information can be accessed using the command:</p>

<pre><code>vArmyKnife help LICENSES
</code></pre>

<p>And can also found in the distributed source code in:</p>

<pre><code>src/main/resources/
</code></pre>

</section>



</div>

		<footer> <small>
		<p>Project maintained by <a href="https://github.com/hartleys/QoRTs">Stephen Hartley</a></p>
		<p>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a> (modified)</p>
		</small> </footer>

</body></html>
